File: testdata/precision-suite-kotlin/clean_safe_path.kt
// Clean: the user-controlled value is reduced to its last path component with
// FilenameUtils.getName(input) before being opened, stripping any `../` traversal
// sequence, so the read is confined to the intended directory. getName is a pure
// sanitizer call (it does not itself open a file), so the sanitized value that
// reaches FileInputStream carries no path-traversal taint. A correct scanner emits
// nothing; any TAINT-004 finding here is a false positive.
package com.example

import java.io.FileInputStream
import org.apache.commons.io.FilenameUtils
import javax.servlet.http.HttpServletRequest

class Downloader {
    fun read(request: HttpServletRequest, baseDir: String) {
        val requested = request.getParameter("file")
        val safe = FilenameUtils.getName(requested)
        val stream = FileInputStream(baseDir + "/" + safe)
        stream.close()
    }
}

File: testdata/precision-suite-rust/clean_safe_path.rs
// Clean: the user-controlled value is reduced to its final path component with
// Path::file_name() before being read, stripping any `../` traversal, so this
// is safe. A correct scanner emits nothing; a TAINT-004 finding here is a false
// positive.
use std::env;
use std::fs;
use std::path::Path;

fn serve_file() -> Vec<u8> {
    let user_path = env::var("FILE").unwrap_or_default();
    // file_name() drops all directory components — ../ can no longer escape.
    let name = Path::new(&user_path).file_name().unwrap_or_default();
    let safe = Path::new("/srv/static").join(name);
    fs::read(safe).unwrap_or_default()
}
